registrations/tasks.py

```diff
@@ -503,6 +536,252 @@ def add_personally_identifiable_fields(registration):
+        """
+        r = requests.get(
+            settings.WASSUP_URL,
+            {
+                'number': settings.WASSUP_NUMBER,
+                'address': address,
+                'wait': True,
+            },
+            headers={
+                'Authorization': "Token {}".format(settings.WASSUP_TOKEN),
+
+            },
+        )
+        r.raise_for_status()
+        data = r.json()
+        existing = filter(lambda d: d.get('exists', False), data.values())
+        return any(existing)
+
+    def create_pmtct_registration(self, registration, operator):
+        if 'whatsapp' in registration.reg_type:
+            reg_type = 'whatsapp_pmtct_prebirth'
+        else:
+            reg_type = 'pmtct_prebirth'
+        data = {
+            'language': registration.data['language'],
+            'mom_dob': registration.data['mom_dob'],
+            'edd': registration.data['edd'],
+            'operator_id': operator['id'],
+        }
+        Registration.objects.create(
+            reg_type=reg_type, registrant_id=registration.registrant_id,
+            source=registration.source, created_by=registration.created_by,
+            data=data)
+
+    def is_identity_subscribed(self, identity, regex):
+        """
+        Checks to see if the identity is subscribed to the specified
+        messageset. Check is done on the short name of the messageset matching
+        the given regular expression
+        """
+        active_subs = utils.sbm_client.get_subscriptions({
+            'identity': identity['id'],
+            'active': True
+        })['results']
+        messagesets = utils.sbm_client.get_messagesets()['results']
+        messagesets = {
+            ms['id']: ms['short_name'] for ms in messagesets
+        }
+        for sub in active_subs:
+            short_name = messagesets[sub['messageset']]
+            if re.search(regex, short_name):
+                return True
+        return False
+
+    def is_valid_clinic_code(self, code):
+        """
+        Checks to see if the specified clinic code is recognised or not
+        """
+        r = requests.get(
+            urljoin(settings.JEMBI_BASE_URL, 'facilityCheck'),
+            {
+                'criteria': "code:{}".format(code),
+            },
+            auth=(settings.JEMBI_USERNAME, settings.JEMBI_PASSWORD),
+        )
+        r.raise_for_status()
+        return len(r.json().get('rows', [])) != 0
+
+    def run(self, registration_id, **kwargs):
+        registration = Registration.objects.get(id=registration_id)
+        msisdn_registrant = registration.data['msisdn_registrant']
+        registrant = self.get_or_update_identity_by_address(msisdn_registrant)
+        device = self.get_or_update_identity_by_address(
+            registration.data['msisdn_device'])
+        registration.registrant_id = registrant['id']
+
+        # Check for existing subscriptions
+        if self.is_identity_subscribed(registrant, r'prebirth\.hw_full'):
```

Comment:
If this is a pmtct registration wouldn't we need to get past this point?

Reply:
So, the `pmtct` option on this endpoint is not whether this is a prebirth or pmtct registration, it's whether this is a prebirth or a prebirth + pmtct registration. Either way it's creating a prebirth registration, so we need this check to avoid creating duplicate prebirth subscriptions.